x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -0,0 +1,203 @@
+ * deprecations can change from one day to another only if aliases or indices are added.
+ */
+public final class DeprecationRoleDescriptorConsumer implements Consumer<Collection<RoleDescriptor>> {
+
+    private static final String ROLE_PERMISSION_DEPRECATION_STANZA = "Role [%s] contains index privileges covering the [%s] alias but"
+            + " which do not cover some of the indices that it points to [%s]. Granting privileges over an alias and hence granting"
+            + " privileges over all the indices that the alias points to is deprecated and will be removed in a future version of"
+            + " Elasticsearch. Instead define permissions exclusively on index names or index name patterns.";
+
+    private static final Logger logger = LogManager.getLogger(DeprecationRoleDescriptorConsumer.class);
+
+    private final DeprecationLogger deprecationLogger;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    
+    // synchronized by this
+    private final Set<String> dailyRoleCache;
+    private final Queue<Collection<RoleDescriptor>> workQueue;
+    private volatile boolean workerBusy;
+
+    public DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool) {
+        this(clusterService, threadPool, new DeprecationLogger(logger));
+    }
+
+    // package-private for testing
+    DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
+        this.deprecationLogger = deprecationLogger;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        // this String Set keeps "<date>-<role>" pairs so that we only log a role once a day.
+        this.dailyRoleCache = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
+                return false == eldest.getKey().startsWith(todayISODate());
+            }
+        });
+        this.workQueue = new LinkedList<>();
+        this.workerBusy = false;
+    }
+
+    @Override
+    public synchronized void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
+        final Collection<RoleDescriptor> uncachedEffectiveRoleDescriptors = new ArrayList<>();
+        for (RoleDescriptor roleDescriptor : effectiveRoleDescriptors) {
+            if (dailyRoleCache.add(buildCacheKey(roleDescriptor))) {
+                uncachedEffectiveRoleDescriptors.add(roleDescriptor);
+            }
+        }
+        if (false == uncachedEffectiveRoleDescriptors.isEmpty()) {
+            workQueue.add(uncachedEffectiveRoleDescriptors);
+            if (false == workerBusy) {
+                workerBusy = true;
+                try {
+                    // spawn another worker on the generic thread pool
+                    threadPool.generic().execute(() -> {
```

Comment:
Can you use a `AbstractRunnable` here? This way you can ensure that workerBusy is set to false even in case of exceptions/rejections in one place, the `onAfter` method?

Reply:
That's a good idea, I implemented as you suggested.